### PR TITLE
feat: server: add clusterconf to specify a cluster it serves for.

### PR DESCRIPTION
**Pull request by drmingdrmer**
### feat: server: add clusterconf to specify a cluster it serves for.

-   feat: CeleServer as server level error type.
    TODO: take a better name for it.

-   feat: Server should be able to be cloned and shared across
    coroutines. Thus core data is put into a Arc<ServerData> struct.

-   feat: Server::new() takes a parameter Storage.

-   feat: Server initialize cluster info, replcias etc.

-   refactor: extract protocol impl out of server. Adds RedisApi
    struct.

## Type of change       

- **New feature**       
- **Refactoring**

# Checklist:

- [x] **Self-review**: I have performed a **self-review** of my own code
- [x] **Comment**:     I have **commented my code**, particularly in hard-to-understand areas
- [x] **Doc**:         I have made corresponding changes to the **documentation**
- [ ] **No-warnings**: My changes generate **no new warnings**
- [ ] **Add-test**:    I have added **tests** that prove my fix is effective or that my feature works
- [x] **Pass**:        New and existing **unit tests pass** locally with my changes

### CODE REVIEW

File: src/bin/cele.rs [hunk 1]
```diff
@@ -16,28 +19,97 @@ use tokio::task::JoinHandle;
...
 use epaxos::conf::Node;
 use epaxos::conf::NodeId;
-use epaxos::qpaxos::*;
+use epaxos::qpaxos::Command;
+use epaxos::qpaxos::Instance;
+use epaxos::qpaxos::InstanceId;
+use epaxos::qpaxos::MyQPaxos;
+use epaxos::qpaxos::OpCode;
+use epaxos::qpaxos::QPaxosServer;
+use epaxos::qpaxos::ReplicaID;
```
Comment (wenbobuaa) [inline src/bin/cele.rs:31]: 为啥没考虑 `use epaxos::qpaxos::{Command, Instance...} ` 这种写法呢，之前这么写的

File: src/bin/cele.rs [hunk 2]
```diff
@@ -16,28 +19,97 @@ use tokio::task::JoinHandle;
...
+    fn from(e: CeleError) -> Response {
+        match e {
+            CeleError::NoGroupForKey(k) => Response::Error(format!("No gruop serves: {}", k)),
+            CeleError::NoLocalReplicaForKey(k) => Response::Error(format!("No replica serve: {}", k)),
+        }
+    }
+}
+
+/// ServerData is shared between threads or coroutine.
+/// TODO: Storage does not need to be shared with Arc any more.
```
Comment (wenbobuaa) [inline src/bin/cele.rs:62]: 如果不用 Arc, Storage 在各个 replica 之间怎么共享呢

File: src/bin/cele.rs (hunk 1 above)
Comment (drmingdrmer) [reply src/bin/cele.rs:31]: rustfmt会把它们合并成一行有时. 然后import修改时就不太容易看出来改了啥了.

File: src/bin/cele.rs (hunk 2 above)
Comment (drmingdrmer) [reply src/bin/cele.rs:62]: 因为引用Storage的Replica是Arc的了, 所以Storage就不用Arc了

**Pull request merged**